src/mysky.ts

```diff
@@ -374,18 +463,241 @@ export class MySky {
+      logout: this.logout.bind(this),
+      portalLogin: this.portalLogin.bind(this),
+      signMessage: this.signMessage.bind(this),
+      signRegistryEntry: this.signRegistryEntry.bind(this),
+      signEncryptedRegistryEntry: this.signEncryptedRegistryEntry.bind(this),
+      userID: this.userID.bind(this),
+      verifyMessageSignature: this.verifyMessageSignature.bind(this),
+    };
+
+    // Enable communication with connector in parent skapp.
+
+    log("Performing handshake");
+    const messenger = new WindowMessenger({
+      localWindow: window,
+      remoteWindow: window.parent,
+      remoteOrigin: "*",
+    });
+    this.parentConnection = ChildHandshake(messenger, methods);
+  }
+
+  /**
+   * Connects to a portal account by either registering or logging in to an
+   * existing account. The resulting cookie will be set on the MySky domain.
+   *
+   * NOTE: We will register "auto re-login" in a separate function.
+   *
+   * @param seed - The user seed.
+   * @param email - The user email.
+   */
+  protected async connectToPortalAccount(seed: Uint8Array, email: string): Promise<void> {
+    log("Entered connectToPortalAccount");
+
+    // Try to connect to the portal account and set the JWT cookie.
+    //
+    // Make requests to login and register in parallel. At most one can succeed,
+    // and this saves a lot of time.
+    try {
+      await Promise.any([register(this.client, seed, email), login(this.client, seed, email)]);
+    } catch (err) {
+      const errors = (err as AggregateError).errors;
+      throw new Error(`Could not register or login: [${errors}]`);
+    }
+  }
+
+  /**
+   * Logs in from MySky UI.
+   *
+   * Flow:
+   *
+   * 0. Unload the permissions provider and seed if stored. (Done in
+   * `setupStorageEventListener`.)
+   *
+   * 1. Always use siasky.net first.
+   *
+   * 2. Get the preferred portal and switch to it (`setPortal`), or if not found
+   * switch to current portal.
+   *
+   * 3. If we got the email, then we register/login to set the JWT cookie
+   * (`connectToPortalAccount`).
+   *
+   * 4. If the email is set, it should set up automatic re-login on JWT
+   * cookie expiry.
+   *
+   * 5. Save the email in user settings.
+   *
+   * 6. Trigger a load of the permissions provider.
+   *
+   * @param seed - The user seed.
+   */
+  protected async loginFromUi(seed: Uint8Array): Promise<void> {
+    log("Entered loginFromUi");
+
+    // Connect to siasky.net first.
+    //
+    // NOTE: Don't use the stored preferred portal here because we are just
+    // logging into a new account and need to get the user settings for the
+    // first time. Always use siasky.net.
+    this.client = getLoginClient(seed, null);
+
+    // Login to portal.
+    {
+      // Get the preferred portal and switch to it.
+      const { portal: preferredPortal, email } = await getUserSettings(this.client, seed, this.mySkyDomain);
+      let storedEmail = email;
```

Comment:
```suggestion
      const { portal: preferredPortal, email: storedEmail } = await getUserSettings(this.client, seed, this.mySkyDomain);
```

Reply:
Making both of them mutable results in a lint error, since one of them is never mutated.

Reply:
Ha ha, ok, it's settled then. Let's keep it as it is.